api/client/swarm/init.go

```diff
@@ -45,19 +54,61 @@ func newInitCommand(dockerCli *client.DockerCli) *cobra.Command {
 func runInit(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts initOptions) error {
 	client := dockerCli.Client()
 	ctx := context.Background()
+	spec := opts.swarmOptions.ToSpec()
+
+	var secret string
+	// If no secret was specified, we create a random one
+	if !flags.Changed("secret") {
+		var secretBytes [generatedSecretEntropyBytes]byte
+
+		if _, err := cryptorand.Read(secretBytes[:]); err != nil {
+			panic(fmt.Errorf("failed to read random bytes: %v", err))
+		}
+
+		var nn big.Int
+		nn.SetBytes(secretBytes[:])
+		secret = fmt.Sprintf("%0[1]*s", maxGeneratedSecretLength, nn.Text(generatedSecretBase))
+
+		fmt.Fprintf(dockerCli.Out(), "No node joining secret provided. Generated random secret:\n\t%s\n\n", secret)
+
+		for i := range spec.AcceptancePolicy.Policies {
+			spec.AcceptancePolicy.Policies[i].Secret = &secret
+		}
+	} else {
+		secret = opts.secret
+	}
```

Comment:
Neither are default values, which also end up in `opts` fields. I think that's ok. This is just another way of setting a default value.